docs/basic/troubleshooting/types.md

```diff
@@ -367,3 +367,28 @@ let baz2: SubIsntType2 = {
 
 - TS also ships with a `Parameters` utility type for extracting the parameters of a function
 - for anything more "custom", the `infer` keyword is the basic building block for this, but takes a bit of getting used to. Look at the source code for the above utility types, and [this example](https://twitter.com/mgechev/status/1211030455224422401?s=20) to get the idea. Basarat [also has a good video on `infer`](https://www.youtube.com/watch?v=ijK-1R-LFII&list=PLYvdvJlnTOjF6aJsWWAt7kZRJvzw-en8B&index=3&t=0s).
+
+## The Types I need doesn't exist!
+
+What's more annoying than modules with unexported types? Modules that are **untyped**!
+
+Fret not! It's just a simple two step process.
+* Create a new type declaration file, say `typedec.d.ts`– if you don't already have one. Ensure that the path to file is resolvable by Typescript. To do so, check the `include` array in the `tsconfig.json` file at the root of your directory.
+```
+// inside tsconfig.json
+{
+...
+    "include": [
+        "src" // automatically resolves if path to declaration is src/typedec.d.ts
+    ],
+...
+}
+```
+* Add the `declare` syntax for your desired module, say `my-untyped-module`– to the declaration file.
+```ts
+// inside typedec.d.ts
+declare module "my-untyped-module"
```

Comment:
TIL that you can just do this without braces! Does it resolve every import to `any`?

Reply:
yup

Reply:
@ferdaber you can easily do away with every untyped import by `declare module '*'`. Saves many lines of declare code but too hacky for my taste.

Reply:
> yup. its a stopgap; i'd like for us to go a liiiitle bit beyond just this

Most of the hooks I use are either typed or custom made. So, I haven't encountered this problem yet.. I'll try to come up with some solution. Coz, it is gonna happen someday and I want to be ready for it.

Reply:
yup. ive done it 3-4 times but had to look it up every time. i shouldve taken the hint and put it in the cheatsheet!

Reply:
> yup. ive done it 3-4 times but had to look it up every time. i shouldve taken the hint and put it in the cheatsheet!

It'll be great if you can share the resources you've found for typing untyped hooks.. I already added a good post for typing class components as reference. Meanwhile, I'll try to find some suggestions and try out a reliable way to type hooks.

Reply:
@sw-yx , After committing the suggested changes, the problem still persists. Am I missing something?

Reply:
honestly idk whats the problem. just ignore it for now, lets focus on content, i can fix the prettier errors later

Reply:
@sw-yx, Sorry for long delay. Looking through custom hook implementations in Github, I found [this hook](https://github.com/donavon/use-dark-mode) which is built in pure JS but exports types for use in TypeScript.

For example, the following parameters for the hook:
```js
const useDarkMode = (
  initialValue = false,
  // The destructured config object
  { 
    element,
    classNameDark,
    classNameLight,
    onChange,
    storageKey = 'darkMode',
    storageProvider,
    global,
  } = {}
) => { ... }
```
are exported in `types/index.d.ts` as:

```ts
declare module 'use-dark-mode' {
  export interface DarkModeConfig {
    classNameDark?: string; 
    classNameLight?: string; 
    element?: HTMLElement; 
    onChange?: (val?: boolean) => void; 
    storageKey?: string; 
    storageProvider?: WindowLocalStorage; 
    global?: Window; 
  }

 // This is an object returned from a call to `useDarkMode`.
  export interface DarkMode {  
    readonly value: boolean; // possible generalisation
    enable: () => void; 
    disable: () => void;
    toggle: () => void;
  }

  export default function useDarkMode(
    initialState?: boolean,  // possible generalisation
    config?: DarkModeConfig
  ): DarkMode;
}
```
I believe the same can be generalised to suit our custom hook typing requirements.

What are your thoughts on this?

Reply:
hmm! you're right. hooks are easier to type than components, haha. i think we can add a Component example to this and then we're done. for example, here is what i had to do to create types for React Router 6 https://gist.github.com/sw-yx/37a6a3d248c2d4031801f0d568904df8

Reply:
Woah, awesome stuff. We can add the lines 11-26 from your gist as an example for creating types for class components as a quick reference. This will also gel well with the previously stated troubleshooting steps.

Reply:
So, reviewing all that's been discussed, the hierarchical focus of topics in the troubleshooting section for this would be:
### Working with Untyped Libraries
   1. **The Lazy Way**
        * Using `declare module "*"`
   2. **The Hard Way**
        1. With *Hooks*
             * `useDarkMode()` example
        2. With *Components*
             * `RR6` example [11-26]

@sw-yx, Is this right or am I missing anything?

Reply:
yes sounds good!

Reply:
@sw-yx I've updated the docs and readme with the stuff we discussed, can you check and help solve the formatting problems?

Reply:
awesome! thank you so much, this works!! :) :) :)